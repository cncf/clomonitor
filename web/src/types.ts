export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
