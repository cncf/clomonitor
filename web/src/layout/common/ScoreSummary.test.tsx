import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { ScoreKind, ScoreType } from '../../types';
import ScoreSummary from './ScoreSummary';

const defaultProps = {
  score: {
    [ScoreType.BestPractices]: 95,
    [ScoreType.Documentation]: 85,
    [ScoreType.Global]: 65,
    [ScoreType.License]: 80,
    score_kind: ScoreKind.Primary,
    [ScoreType.Security]: 0,
  },
  bigSize: false,
};

describe('ScoreSummary', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <ScoreSummary {...defaultProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <Router>
          <ScoreSummary {...defaultProps} />
        </Router>
      );

      expect(screen.getByText('Documentation')).toBeInTheDocument();
      expect(screen.getByText('License')).toBeInTheDocument();
      expect(screen.getByText('Best Practices')).toBeInTheDocument();
      expect(screen.getByText('Security')).toBeInTheDocument();

      expect(screen.getByTestId('global-score')).toBeInTheDocument();
      expect(screen.getAllByTestId('line')).toHaveLength(4);
      expect(screen.getAllByTestId('peak')).toHaveLength(4);
    });

    it('renders correct classes when bigSize is true', () => {
      const { container } = render(
        <Router>
          <ScoreSummary {...defaultProps} bigSize />
        </Router>
      );

      expect(container.children[0]).toHaveClass('bigSize');
      expect(container.children[0].children[1]).toHaveClass('px-0 px-sm-3');
      expect(container.children[0].children[1].children[0]).toHaveClass('gx-4 gx-md-5');
      expect(screen.getByText('85')).toHaveClass('bigSize');
      expect(screen.getByTestId('global-score').parentNode).toHaveClass('mx-3');
    });
  });
});
