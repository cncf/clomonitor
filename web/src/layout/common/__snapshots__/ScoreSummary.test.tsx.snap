// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreSummary creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="align-items-center d-flex flex-column flex-md-row summary flex-lg-column flex-xl-row"
  >
    <div
      class="d-none d-md-block d-lg-none d-xl-block d-lg-none d-xl-block"
    >
      <div
        class="d-flex align-items-center justify-content-center rounded-pill value global me-0 me-sm-4 mb-2 mb-sm-0"
        data-testid="global-score"
      >
        65
      </div>
    </div>
    <div
      class="flex-grow-1 w-100 position-relative categories"
    >
      <div
        class="row"
      >
        <div
          class="col-6 text-truncate"
        >
          <div
            class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
          >
            <span
              class="pe-1 d-inline-block position-relative icon"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </span>
            <span
              class="text-truncate d-none d-md-inline-block"
            >
              Documentation
            </span>
            <span
              class="d-inline-block d-md-none text-truncate"
            >
              Docs
            </span>
          </div>
          <div
            class="d-flex flex-row bg-white position-relative border overflow-hidden line"
          >
            <div
              class="text-center fw-bold font-monospace value "
            >
              85
            </div>
            <div
              class="flex-grow-1 w-100 position-relative"
            >
              <div
                class="position-absolute top-0 start-0 bottom-0"
                data-testid="line"
                style="width: calc(85% - 5px);"
              >
                <div
                  class="position-absolute arrow "
                  data-testid="peak"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-6 text-truncate"
        >
          <div
            class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
          >
            <span
              class="pe-1 d-inline-block position-relative icon"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 14 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 4c-.83 0-1.5-.67-1.5-1.5S6.17 1 7 1s1.5.67 1.5 1.5S7.83 4 7 4zm7 6c0 1.11-.89 2-2 2h-1c-1.11 0-2-.89-2-2l2-4h-1c-.55 0-1-.45-1-1H8v8c.42 0 1 .45 1 1h1c.42 0 1 .45 1 1H3c0-.55.58-1 1-1h1c0-.55.58-1 1-1h.03L6 5H5c0 .55-.45 1-1 1H3l2 4c0 1.11-.89 2-2 2H2c-1.11 0-2-.89-2-2l2-4H1V5h3c0-.55.45-1 1-1h4c.55 0 1 .45 1 1h3v1h-1l2 4zM2.5 7L1 10h3L2.5 7zM13 10l-1.5-3-1.5 3h3z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="text-truncate"
            >
              License
            </span>
          </div>
          <div
            class="d-flex flex-row bg-white position-relative border overflow-hidden line"
          >
            <div
              class="text-center fw-bold font-monospace value "
            >
              80
            </div>
            <div
              class="flex-grow-1 w-100 position-relative"
            >
              <div
                class="position-absolute top-0 start-0 bottom-0"
                data-testid="line"
                style="width: calc(80% - 5px);"
              >
                <div
                  class="position-absolute arrow "
                  data-testid="peak"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row mt-2"
      >
        <div
          class="col-6 text-truncate"
        >
          <div
            class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
          >
            <span
              class="pe-1 d-inline-block position-relative icon"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 32c-70.7 0-128 57.3-128 128s57.3 128 128 128 128-57.3 128-128S326.7 32 256 32zm0 208c-44.2 0-80-35.8-80-80s35.8-80 80-80 80 35.8 80 80-35.8 80-80 80zM193.7 307.4c-19.1-8.1-36.2-19.6-50.8-34.3-1.4-1.4-2.8-2.8-4.1-4.3L64 400h96l48 80 48-105.8 25.5-56.2c-8.4 1.3-16.9 2-25.5 2-21.6 0-42.5-4.2-62.3-12.6zM373.3 268.9c-1.3 1.4-2.7 2.9-4.1 4.3-14.6 14.6-31.7 26.2-50.7 34.2L294 361.2l-21.9 48.4L304 480l48-80h96l-74.7-131.1z"
                />
              </svg>
            </span>
            <span
              class="text-truncate"
            >
              Best Practices
            </span>
          </div>
          <div
            class="d-flex flex-row bg-white position-relative border overflow-hidden line"
          >
            <div
              class="text-center fw-bold font-monospace value "
            >
              95
            </div>
            <div
              class="flex-grow-1 w-100 position-relative"
            >
              <div
                class="position-absolute top-0 start-0 bottom-0"
                data-testid="line"
                style="width: calc(95% - 5px);"
              >
                <div
                  class="position-absolute arrow "
                  data-testid="peak"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-6 text-truncate"
        >
          <div
            class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
          >
            <span
              class="pe-1 d-inline-block position-relative icon"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2C9.243 2 7 4.243 7 7v2H6c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2h-1V7c0-2.757-2.243-5-5-5zM9 7c0-1.654 1.346-3 3-3s3 1.346 3 3v2H9V7zm9.002 13H13v-2.278c.595-.347 1-.985 1-1.722 0-1.103-.897-2-2-2s-2 .897-2 2c0 .736.405 1.375 1 1.722V20H6v-9h12l.002 9z"
                />
              </svg>
            </span>
            <span
              class="text-truncate"
            >
              Security
            </span>
          </div>
          <div
            class="d-flex flex-row bg-white position-relative border overflow-hidden line"
          >
            <div
              class="text-center fw-bold font-monospace value "
            >
              0
            </div>
            <div
              class="flex-grow-1 w-100 position-relative"
            >
              <div
                class="position-absolute top-0 start-0 bottom-0"
                data-testid="line"
                style="width: calc(0% - 5px);"
              >
                <div
                  class="position-absolute arrow "
                  data-testid="peak"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
